src/Storages/MergeTree/DataPartsExchange.cpp

```diff
@@ -351,22 +349,59 @@ MergeTreeData::DataPart::Checksums Service::sendPartFromDisk(
     return data_checksums;
 }
 
+bool wait_loop(UInt32 wait_timeout_ms, const std::function<bool()> & pred)
+{
+    static const UInt32 loop_delay_ms = 5;
+
+    /// this is sleep-based wait, it has to be short
+    chassert(wait_timeout_ms < 2000);
+
+    if (pred())
+        return true;
+
+    Stopwatch timer;
+    sleepForMilliseconds(loop_delay_ms);
+    while (!pred() && timer.elapsedMilliseconds() < wait_timeout_ms)
+    {
+        sleepForMilliseconds(loop_delay_ms);
+    }
+
+    return pred();
+}
+
 MergeTreeData::DataPartPtr Service::findPart(const String & name)
 {
     /// It is important to include Outdated parts here because remote replicas cannot reliably
     /// determine the local state of the part, so queries for the parts in these states are completely normal.
     MergeTreeData::DataPartPtr part;
 
-    /// Ephemeral zero-copy lock may be lost for PreActive parts
+    part = data.getPartIfExists(name, {MergeTreeDataPartState::PreActive, MergeTreeDataPartState::Active, MergeTreeDataPartState::Outdated});
+
+    if (!part)
+        throw Exception(ErrorCodes::NO_SUCH_DATA_PART, "No part {} in table", name);
+
     bool zero_copy_enabled = data.getSettings()->allow_remote_fs_zero_copy_replication;
-    if (zero_copy_enabled)
-        part = data.getPartIfExists(name, {MergeTreeDataPartState::Active, MergeTreeDataPartState::Outdated});
-    else
-        part = data.getPartIfExists(name, {MergeTreeDataPartState::PreActive, MergeTreeDataPartState::Active, MergeTreeDataPartState::Outdated});
-    if (part)
+    if (!zero_copy_enabled)
         return part;
 
-    throw Exception(ErrorCodes::NO_SUCH_DATA_PART, "No part {} in table", name);
+    /// Ephemeral zero-copy lock may be lost for PreActive parts
+    /// do not expose PreActive parts for zero-copy
+
+    static const UInt32 wait_timeout_ms = 1000;
+    auto pred = [&] ()
+    {
+        auto lock = data.lockParts();
+        return part->getState() != MergeTreeDataPartState::PreActive;
+    };
+
+    bool pred_result = wait_loop(wait_timeout_ms, pred);
+    if (!pred_result)
+        throw Exception(
+                ErrorCodes::ABORTED,
+                "Part {} is in PreActive state for {} ms. Another host has to be asked.",
```

Comment:
Could we make it even more obvious for the client?

Something like `Could not exchange part {} as it's in preActive state ({} ms) and it uses zero copy replication. This is expected behaviour and the client will retry fetching the part automatically`.

That way we avoid (more) questions about the error.